diff --git a/external-crates/move/crates/move-binary-format/src/lib.rs b/external-crates/move/crates/move-binary-format/src/lib.rs
index a0ebacf7941f2..1c8d4529a79f6 100644
--- a/external-crates/move/crates/move-binary-format/src/lib.rs
+++ b/external-crates/move/crates/move-binary-format/src/lib.rs
@@ -158,8 +158,10 @@ macro_rules! safe_unwrap_err {
         match $e {
             Ok(x) => x,
             Err(e) => {
-                let err = PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                    .with_message(format!("{}:{} {:#}", file!(), line!(), e));
+                let err = move_binary_format::errors::PartialVMError::new(
+                    move_core_types::vm_status::StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                )
+                .with_message(format!("{}:{} {:#}", file!(), line!(), e));
                 if cfg!(debug_assertions) {
                     panic!("{:?}", err);
                 } else {
@@ -188,6 +190,32 @@ macro_rules! safe_assert {
     }};
 }
 
+/// Similar as above, but asserts two expressions are equal.
+#[macro_export]
+macro_rules! safe_assert_eq {
+    ($left:expr, $right:expr) => {{
+        let left = &$left;
+        let right = &$right;
+        if left != right {
+            let err = $crate::errors::PartialVMError::new(
+                move_core_types::vm_status::StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+            )
+            .with_message(format!(
+                "{}:{} (assert_eq: {:?} != {:?})",
+                file!(),
+                line!(),
+                left,
+                right,
+            ));
+            if cfg!(debug_assertions) {
+                panic!("{:?}", err)
+            } else {
+                return Err(err);
+            }
+        }
+    }};
+}
+
 /// Create a PartialVMError with the given error code and an optional message.
 #[macro_export]
 macro_rules! partial_vm_error {
diff --git a/sui-execution/latest/sui-move-natives/src/accumulator.rs b/sui-execution/latest/sui-move-natives/src/accumulator.rs
index 72233a3d095e0..732e7bb1a4576 100644
--- a/sui-execution/latest/sui-move-natives/src/accumulator.rs
+++ b/sui-execution/latest/sui-move-natives/src/accumulator.rs
@@ -6,6 +6,7 @@ use crate::{
     object_runtime::{MoveAccumulatorAction, MoveAccumulatorValue, ObjectRuntime},
 };
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::{safe_unwrap, safe_unwrap_err};
 use move_core_types::account_address::AccountAddress;
 use move_vm_runtime::execution::{Type, values::Value};
 use move_vm_runtime::native_charge_gas_early_exit;
@@ -46,20 +47,12 @@ fn emit_event(
     // TODO(address-balances): add specific cost for this
     native_charge_gas_early_exit!(context, event_emit_cost_params.event_emit_cost_base);
 
-    let amount = args.pop_back().unwrap().value_as::<u64>().unwrap();
-    let recipient = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap();
-    let accumulator: ObjectID = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
-
-    let ty_tag = context.type_to_type_tag(&ty_args.pop().unwrap())?;
+    let amount = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<u64>());
+    let recipient = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>());
+    let accumulator: ObjectID =
+        safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>()).into();
+
+    let ty_tag = context.type_to_type_tag(&safe_unwrap!(ty_args.pop()))?;
 
     let obj_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
 
@@ -82,8 +75,8 @@ pub fn record_settlement_sui_conservation(
     debug_assert!(ty_args.is_empty());
     debug_assert!(args.len() == 2);
 
-    let output_sui = args.pop_back().unwrap().value_as::<u64>().unwrap();
-    let input_sui = args.pop_back().unwrap().value_as::<u64>().unwrap();
+    let output_sui = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<u64>());
+    let input_sui = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<u64>());
 
     let obj_runtime: &mut ObjectRuntime = context.extensions_mut().get_mut()?;
 
diff --git a/sui-execution/latest/sui-move-natives/src/address.rs b/sui-execution/latest/sui-move-natives/src/address.rs
index 1361fcfa77bbc..0362a46eb4a63 100644
--- a/sui-execution/latest/sui-move-natives/src/address.rs
+++ b/sui-execution/latest/sui-move-natives/src/address.rs
@@ -3,6 +3,7 @@
 
 use crate::{NativesCostTable, get_extension};
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::{safe_unwrap, safe_unwrap_err};
 use move_core_types::{account_address::AccountAddress, gas_algebra::InternalGas, u256::U256};
 use move_vm_runtime::native_charge_gas_early_exit;
 use move_vm_runtime::natives::functions::NativeContext;
@@ -84,8 +85,9 @@ pub fn to_u256(
     let mut addr_bytes_le = addr.to_vec();
     addr_bytes_le.reverse();
 
-    // unwrap safe because we know addr_bytes_le is length 32
-    let u256_val = Value::u256(U256::from_le_bytes(&addr_bytes_le.try_into().unwrap()));
+    let u256_val = Value::u256(U256::from_le_bytes(&safe_unwrap!(
+        addr_bytes_le.try_into().ok()
+    )));
     Ok(NativeResult::ok(context.gas_used(), smallvec![u256_val]))
 }
 
@@ -121,7 +123,8 @@ pub fn from_u256(
     let mut u256_bytes = u256.to_le_bytes().to_vec();
     u256_bytes.reverse();
 
-    // unwrap safe because we are passing a 32 byte slice
-    let addr_val = Value::address(AccountAddress::from_bytes(&u256_bytes[..]).unwrap());
+    let addr_val = Value::address(safe_unwrap_err!(AccountAddress::from_bytes(
+        &u256_bytes[..]
+    )));
     Ok(NativeResult::ok(context.gas_used(), smallvec![addr_val]))
 }
diff --git a/sui-execution/latest/sui-move-natives/src/config.rs b/sui-execution/latest/sui-move-natives/src/config.rs
index d689ba6b0e599..18778c321aa7a 100644
--- a/sui-execution/latest/sui-move-natives/src/config.rs
+++ b/sui-execution/latest/sui-move-natives/src/config.rs
@@ -6,6 +6,7 @@ use crate::{
     object_runtime::ObjectRuntime,
 };
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{safe_assert_eq, safe_unwrap};
 use move_core_types::{
     account_address::AccountAddress, gas_algebra::InternalGas, language_storage::StructTag,
     runtime_value as R, vm_status::StatusCode,
@@ -37,8 +38,8 @@ pub fn read_setting_impl(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert_eq!(ty_args.len(), 4);
-    assert_eq!(args.len(), 3);
+    safe_assert_eq!(ty_args.len(), 4);
+    safe_assert_eq!(args.len(), 3);
 
     let ConfigReadSettingImplCostParams {
         config_read_setting_impl_cost_base,
@@ -60,10 +61,10 @@ pub fn read_setting_impl(
     // Charge base fee
     native_charge_gas_early_exit!(context, config_read_setting_impl_cost_base);
 
-    let value_ty = ty_args.pop().unwrap();
-    let setting_data_value_ty = ty_args.pop().unwrap();
-    let setting_value_ty = ty_args.pop().unwrap();
-    let field_setting_ty = ty_args.pop().unwrap();
+    let value_ty = safe_unwrap!(ty_args.pop());
+    let setting_data_value_ty = safe_unwrap!(ty_args.pop());
+    let setting_value_ty = safe_unwrap!(ty_args.pop());
+    let field_setting_ty = safe_unwrap!(ty_args.pop());
 
     let current_epoch = pop_arg!(args, u64);
     let name_df_addr = pop_arg!(args, AccountAddress);
diff --git a/sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs b/sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs
index 30ac85d010dc0..dfe5735165373 100644
--- a/sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs
+++ b/sui-execution/latest/sui-move-natives/src/crypto/group_ops.rs
@@ -9,6 +9,7 @@ use fastcrypto::groups::{
 };
 use fastcrypto::serde_helpers::ToFromByteArray;
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::safe_unwrap_err;
 use move_core_types::gas_algebra::InternalGas;
 use move_core_types::vm_status::StatusCode;
 use move_vm_runtime::native_charge_gas_early_exit;
@@ -765,8 +766,7 @@ where
                 .map(|per_addition| base + per_addition * num_of_additions.into()))
         );
 
-        let r = G::multi_scalar_mul(&scalars, &points)
-            .expect("Already checked the lengths of the vectors");
+        let r = safe_unwrap_err!(G::multi_scalar_mul(&scalars, &points));
         Ok(NativeResult::ok(
             context.gas_used(),
             smallvec![Value::vector_u8(r.to_byte_array().to_vec())],
diff --git a/sui-execution/latest/sui-move-natives/src/crypto/hmac.rs b/sui-execution/latest/sui-move-natives/src/crypto/hmac.rs
index dce016fd9c227..5b11db63da92b 100644
--- a/sui-execution/latest/sui-move-natives/src/crypto/hmac.rs
+++ b/sui-execution/latest/sui-move-natives/src/crypto/hmac.rs
@@ -3,6 +3,7 @@
 use crate::{NativesCostTable, get_extension};
 use fastcrypto::{hmac, traits::ToFromBytes};
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::safe_unwrap_err;
 use move_core_types::gas_algebra::InternalGas;
 use move_vm_runtime::{
     execution::{
@@ -72,8 +73,7 @@ pub fn hmac_sha3_256(
                     .into()
     );
 
-    let hmac_key = hmac::HmacKey::from_bytes(&key.as_bytes_ref()?)
-        .expect("HMAC key can be of any length and from_bytes should always succeed");
+    let hmac_key = safe_unwrap_err!(hmac::HmacKey::from_bytes(&key.as_bytes_ref()?));
     let cost = context.gas_used();
 
     Ok(NativeResult::ok(
diff --git a/sui-execution/latest/sui-move-natives/src/dynamic_field.rs b/sui-execution/latest/sui-move-natives/src/dynamic_field.rs
index bac9b03bc265b..b1fa1eb8397f2 100644
--- a/sui-execution/latest/sui-move-natives/src/dynamic_field.rs
+++ b/sui-execution/latest/sui-move-natives/src/dynamic_field.rs
@@ -10,6 +10,9 @@ use crate::{
     },
 };
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{
+    partial_vm_error, safe_assert, safe_assert_eq, safe_unwrap, safe_unwrap_err,
+};
 use move_core_types::{
     account_address::AccountAddress,
     gas_algebra::InternalGas,
@@ -43,13 +46,13 @@ const BORROW_ABSTRACT_SIZE: u64 = 8;
 
 macro_rules! get_or_fetch_object {
     ($context:ident, $ty_args:ident, $parent:ident, $child_id:ident, $ty_cost_per_byte:expr) => {{
-        let child_ty = $ty_args.pop().unwrap();
+        let child_ty = safe_unwrap!($ty_args.pop());
         native_charge_gas_early_exit!(
             $context,
             $ty_cost_per_byte * u64::from(child_ty.size()?).into()
         );
 
-        assert!($ty_args.is_empty());
+        safe_assert!($ty_args.is_empty());
         let (tag, layout, annotated_layout) = match crate::get_tag_and_layouts($context, &child_ty)?
         {
             Some(res) => res,
@@ -94,8 +97,8 @@ pub fn hash_type_and_key(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert_eq!(ty_args.len(), 1);
-    assert_eq!(args.len(), 2);
+    safe_assert_eq!(ty_args.len(), 1);
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_hash_type_and_key_cost_params = get_extension!(context, NativesCostTable)?
         .dynamic_field_hash_type_and_key_cost_params
@@ -107,8 +110,8 @@ pub fn hash_type_and_key(
         dynamic_field_hash_type_and_key_cost_params.dynamic_field_hash_type_and_key_cost_base
     );
 
-    let k_ty = ty_args.pop().unwrap();
-    let k: Value = args.pop_back().unwrap();
+    let k_ty = safe_unwrap!(ty_args.pop());
+    let k: Value = safe_unwrap!(args.pop_back());
     let parent = pop_arg!(args, AccountAddress);
 
     // Get size info for costing for derivations, serializations, etc
@@ -176,8 +179,8 @@ pub fn add_child_object(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.len() == 1);
-    assert!(args.len() == 2);
+    safe_assert_eq!(ty_args.len(), 1);
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_add_child_object_cost_params = get_extension!(context, NativesCostTable)?
         .dynamic_field_add_child_object_cost_params
@@ -189,9 +192,9 @@ pub fn add_child_object(
         dynamic_field_add_child_object_cost_params.dynamic_field_add_child_object_cost_base
     );
 
-    let child = args.pop_back().unwrap();
+    let child = safe_unwrap!(args.pop_back());
     let parent = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
 
     // The value already exists, the size of the value is irrelevant
     let child_value_size = PRE_EXISTING_ABSTRACT_SIZE;
@@ -204,12 +207,11 @@ pub fn add_child_object(
     );
 
     // TODO remove this copy_value, which will require VM changes
-    let child_id = get_object_id(child.copy_value())
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
-    let child_ty = ty_args.pop().unwrap();
+    let child_id = safe_unwrap_err!(
+        get_object_id(child.copy_value()).and_then(|v| v.value_as::<AccountAddress>())
+    )
+    .into();
+    let child_ty = safe_unwrap!(ty_args.pop());
     let child_type_size = u64::from(child_ty.size()?);
 
     native_charge_gas_early_exit!(
@@ -219,7 +221,7 @@ pub fn add_child_object(
             * child_type_size.into()
     );
 
-    assert!(ty_args.is_empty());
+    safe_assert!(ty_args.is_empty());
     let tag = match context.type_to_type_tag(&child_ty)? {
         TypeTag::Struct(s) => *s,
         _ => {
@@ -272,8 +274,8 @@ pub fn borrow_child_object(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.len() == 1);
-    assert!(args.len() == 2);
+    safe_assert_eq!(ty_args.len(), 1);
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_borrow_child_object_cost_params = get_extension!(context, NativesCostTable)?
         .dynamic_field_borrow_child_object_cost_params
@@ -285,15 +287,14 @@ pub fn borrow_child_object(
 
     let child_id = pop_arg!(args, AccountAddress).into();
 
-    let parent_uid = pop_arg!(args, StructRef).read_ref().unwrap();
+    let parent_uid = safe_unwrap_err!(pop_arg!(args, StructRef).read_ref());
     // UID { id: ID { bytes: address } }
-    let parent = get_nested_struct_field(parent_uid, &[0, 0])
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
+    let parent = safe_unwrap_err!(
+        get_nested_struct_field(parent_uid, &[0, 0]).and_then(|v| v.value_as::<AccountAddress>())
+    )
+    .into();
 
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let global_value_result = get_or_fetch_object!(
         context,
         ty_args,
@@ -311,8 +312,16 @@ pub fn borrow_child_object(
     if !global_value.exists()? {
         return Ok(NativeResult::err(context.gas_used(), E_KEY_DOES_NOT_EXIST));
     }
-    let child_ref = global_value.borrow_global().inspect_err(|err| {
-        assert!(err.major_status() != StatusCode::MISSING_DATA);
+    let child_ref = global_value.borrow_global().map_err(|err| {
+        if err.major_status() == StatusCode::MISSING_DATA {
+            debug_assert!(false);
+            partial_vm_error!(
+                UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                "borrow_global returned MISSING_DATA after exists() was true"
+            )
+        } else {
+            err
+        }
     })?;
 
     charge_cache_or_load_gas!(context, cache_info);
@@ -362,8 +371,8 @@ pub fn remove_child_object(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.len() == 1);
-    assert!(args.len() == 2);
+    safe_assert_eq!(ty_args.len(), 1);
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_remove_child_object_cost_params = get_extension!(context, NativesCostTable)?
         .dynamic_field_remove_child_object_cost_params
@@ -375,7 +384,7 @@ pub fn remove_child_object(
 
     let child_id = pop_arg!(args, AccountAddress).into();
     let parent = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let global_value_result = get_or_fetch_object!(
         context,
         ty_args,
@@ -394,8 +403,16 @@ pub fn remove_child_object(
     if !global_value.exists()? {
         return Ok(NativeResult::err(context.gas_used(), E_KEY_DOES_NOT_EXIST));
     }
-    let child = global_value.move_from().inspect_err(|err| {
-        assert!(err.major_status() != StatusCode::MISSING_DATA);
+    let child = global_value.move_from().map_err(|err| {
+        if err.major_status() == StatusCode::MISSING_DATA {
+            debug_assert!(false);
+            partial_vm_error!(
+                UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                "move_from returned MISSING_DATA after exists() was true"
+            )
+        } else {
+            err
+        }
     })?;
 
     charge_cache_or_load_gas!(context, cache_info);
@@ -440,8 +457,8 @@ pub fn has_child_object(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
-    assert!(args.len() == 2);
+    safe_assert!(ty_args.is_empty());
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_has_child_object_cost_params = get_extension!(context, NativesCostTable)?
         .dynamic_field_has_child_object_cost_params
@@ -481,8 +498,8 @@ pub fn has_child_object_with_ty(
     mut ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.len() == 1);
-    assert!(args.len() == 2);
+    safe_assert_eq!(ty_args.len(), 1);
+    safe_assert_eq!(args.len(), 2);
 
     let dynamic_field_has_child_object_with_ty_cost_params =
         get_extension!(context, NativesCostTable)?
@@ -496,8 +513,8 @@ pub fn has_child_object_with_ty(
 
     let child_id = pop_arg!(args, AccountAddress).into();
     let parent = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
-    let ty = ty_args.pop().unwrap();
+    safe_assert!(args.is_empty());
+    let ty = safe_unwrap!(ty_args.pop());
 
     native_charge_gas_early_exit!(
         context,
diff --git a/sui-execution/latest/sui-move-natives/src/event.rs b/sui-execution/latest/sui-move-natives/src/event.rs
index 9bea76a4468e6..400ab26c8fc6c 100644
--- a/sui-execution/latest/sui-move-natives/src/event.rs
+++ b/sui-execution/latest/sui-move-natives/src/event.rs
@@ -6,6 +6,7 @@ use crate::{
     object_runtime::{MoveAccumulatorAction, MoveAccumulatorValue, ObjectRuntime},
 };
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{safe_assert, safe_assert_eq, safe_unwrap, safe_unwrap_err};
 use move_core_types::{
     account_address::AccountAddress, gas_algebra::InternalGas, language_storage::TypeTag,
     vm_status::StatusCode,
@@ -50,8 +51,8 @@ pub fn emit(
     debug_assert!(ty_args.len() == 1);
     debug_assert!(args.len() == 1);
 
-    let ty = ty_args.pop().unwrap();
-    let event_value = args.pop_back().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
+    let event_value = safe_unwrap!(args.pop_back());
     emit_impl(context, ty, event_value, None)
 }
 
@@ -71,13 +72,13 @@ pub fn emit_authenticated_impl(
         return Ok(NativeResult::err(cost, NOT_SUPPORTED));
     }
 
-    let event_ty = ty_args.pop().unwrap();
+    let event_ty = safe_unwrap!(ty_args.pop());
     // This type is always sui::event::EventStreamHead
-    let stream_head_ty = ty_args.pop().unwrap();
+    let stream_head_ty = safe_unwrap!(ty_args.pop());
 
-    let event_value = args.pop_back().unwrap();
-    let stream_id = args.pop_back().unwrap();
-    let accumulator_id = args.pop_back().unwrap();
+    let event_value = safe_unwrap!(args.pop_back());
+    let stream_id = safe_unwrap!(args.pop_back());
+    let accumulator_id = safe_unwrap!(args.pop_back());
 
     emit_impl(
         context,
@@ -146,9 +147,7 @@ fn emit_impl(
     if stream_ref.is_some() {
         native_charge_gas_early_exit!(
             context,
-            // this code cannot be reached in protocol versions which don't define
-            // event_emit_auth_stream_cost
-            event_emit_cost_params.event_emit_auth_stream_cost.unwrap()
+            safe_unwrap!(event_emit_cost_params.event_emit_auth_stream_cost)
         );
     }
 
@@ -207,8 +206,10 @@ fn emit_impl(
         stream_head_ty: _,
     }) = stream_ref
     {
-        let stream_id_addr: AccountAddress = stream_id.value_as::<AccountAddress>().unwrap();
-        let accumulator_id: ObjectID = accumulator_id.value_as::<AccountAddress>().unwrap().into();
+        let stream_id_addr: AccountAddress =
+            safe_unwrap_err!(stream_id.value_as::<AccountAddress>());
+        let accumulator_id: ObjectID =
+            safe_unwrap_err!(accumulator_id.value_as::<AccountAddress>()).into();
         let event_idx = obj_runtime
             .state
             .total_events_emitted()
@@ -221,7 +222,7 @@ fn emit_impl(
             accumulator_id,
             MoveAccumulatorAction::Merge,
             stream_id_addr,
-            stream_head_type_tag.unwrap(),
+            safe_unwrap!(stream_head_type_tag),
             MoveAccumulatorValue::EventRef(event_idx),
         )?;
     }
@@ -235,8 +236,8 @@ pub fn num_events(
     ty_args: Vec<Type>,
     args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
-    assert!(args.is_empty());
+    safe_assert!(ty_args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime_ref: &ObjectRuntime = get_extension!(context)?;
     let num_events = object_runtime_ref.state.events().len();
     Ok(NativeResult::ok(
@@ -251,10 +252,10 @@ pub fn get_events_by_type(
     mut ty_args: Vec<Type>,
     args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert_eq!(ty_args.len(), 1);
-    let specified_ty = ty_args.pop().unwrap();
+    safe_assert_eq!(ty_args.len(), 1);
+    let specified_ty = safe_unwrap!(ty_args.pop());
     let specialization: VectorSpecialization = (&specified_ty).try_into()?;
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime_ref: &ObjectRuntime = get_extension!(context)?;
     let specified_type_tag = match context.type_to_type_tag(&specified_ty)? {
         TypeTag::Struct(s) => *s,
diff --git a/sui-execution/latest/sui-move-natives/src/funds_accumulator.rs b/sui-execution/latest/sui-move-natives/src/funds_accumulator.rs
index 8e4b37a67c3a0..d4f9b54c276bb 100644
--- a/sui-execution/latest/sui-move-natives/src/funds_accumulator.rs
+++ b/sui-execution/latest/sui-move-natives/src/funds_accumulator.rs
@@ -4,6 +4,7 @@
 use std::collections::VecDeque;
 
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{safe_unwrap, safe_unwrap_err};
 use move_core_types::{account_address::AccountAddress, u256::U256, vm_status::StatusCode};
 use move_vm_runtime::{
     execution::{
@@ -40,9 +41,9 @@ pub fn add_to_accumulator_address(
         .clone();
     native_charge_gas_early_exit!(context, event_emit_cost_params.event_emit_cost_base);
 
-    let ty_tag = context.type_to_type_tag(&ty_args.pop().unwrap())?;
+    let ty_tag = context.type_to_type_tag(&safe_unwrap!(ty_args.pop()))?;
 
-    let Some(value) = args.pop_back().unwrap().value_as::<Struct>().ok() else {
+    let Some(value) = safe_unwrap!(args.pop_back()).value_as::<Struct>().ok() else {
         // TODO in the future this is guaranteed/checked via a custom verifier rule
         debug_assert!(false);
         return Err(
@@ -51,17 +52,9 @@ pub fn add_to_accumulator_address(
             ),
         );
     };
-    let recipient = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap();
-    let accumulator: ObjectID = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
+    let recipient = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>());
+    let accumulator: ObjectID =
+        safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>()).into();
 
     // TODO this will need to look at the layout of T when this is not guaranteed to be a Balance
     let Some([amount]): Option<[Value; 1]> = value.unpack().collect::<Vec<_>>().try_into().ok()
@@ -117,20 +110,12 @@ pub fn withdraw_from_accumulator_address(
         .clone();
     native_charge_gas_early_exit!(context, event_emit_cost_params.event_emit_cost_base);
 
-    let ty_tag = context.type_to_type_tag(&ty_args.pop().unwrap())?;
-
-    let value = args.pop_back().unwrap().value_as::<U256>().unwrap();
-    let recipient = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap();
-    let accumulator: ObjectID = args
-        .pop_back()
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
+    let ty_tag = context.type_to_type_tag(&safe_unwrap!(ty_args.pop()))?;
+
+    let value = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<U256>());
+    let recipient = safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>());
+    let accumulator: ObjectID =
+        safe_unwrap_err!(safe_unwrap!(args.pop_back()).value_as::<AccountAddress>()).into();
 
     // TODO this will need to look at the layout of T when this is not guaranteed to be a Balance
     let Ok(amount): Result<u64, _> = value.try_into() else {
diff --git a/sui-execution/latest/sui-move-natives/src/lib.rs b/sui-execution/latest/sui-move-natives/src/lib.rs
index b738aaaaa7674..608334ec09fbc 100644
--- a/sui-execution/latest/sui-move-natives/src/lib.rs
+++ b/sui-execution/latest/sui-move-natives/src/lib.rs
@@ -49,6 +49,7 @@ use better_any::{Tid, TidAble};
 use crypto::nitro_attestation::{self, NitroAttestationCostParams};
 use crypto::vdf::{self, VDFCostParams};
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::safe_unwrap;
 use move_core_types::{
     annotated_value as A,
     gas_algebra::{AbstractMemorySize, InternalGas},
@@ -1296,6 +1297,7 @@ pub fn all_natives(silent: bool, protocol_config: &ProtocolConfig) -> NativeFunc
             .map(|(module_name, func_name, func)| {
                 (
                     SUI_FRAMEWORK_ADDRESS,
+                    // Safe: string literals are always valid identifiers
                     Identifier::new(module_name).unwrap(),
                     Identifier::new(func_name).unwrap(),
                     func,
@@ -1312,6 +1314,7 @@ pub fn all_natives(silent: bool, protocol_config: &ProtocolConfig) -> NativeFunc
         .map(|(module_name, func_name, func)| {
             (
                 SUI_SYSTEM_ADDRESS,
+                // Safe: string literals are always valid identifiers
                 Identifier::new(module_name).unwrap(),
                 Identifier::new(func_name).unwrap(),
                 func,
@@ -1351,7 +1354,7 @@ pub fn get_nested_struct_field(mut v: Value, offsets: &[usize]) -> Result<Value,
 
 pub fn get_nth_struct_field(v: Value, n: usize) -> Result<Value, PartialVMError> {
     let mut itr = v.value_as::<Struct>()?.unpack();
-    Ok(itr.nth(n).unwrap())
+    Ok(safe_unwrap!(itr.nth(n)))
 }
 
 /// Returns the struct tag, non-annotated type layout, and fully annotated type layout of `ty`.
diff --git a/sui-execution/latest/sui-move-natives/src/test_scenario.rs b/sui-execution/latest/sui-move-natives/src/test_scenario.rs
index 84eed1a21baf4..027e565043668 100644
--- a/sui-execution/latest/sui-move-natives/src/test_scenario.rs
+++ b/sui-execution/latest/sui-move-natives/src/test_scenario.rs
@@ -8,6 +8,7 @@ use crate::{
 use better_any::{Tid, TidAble};
 use indexmap::{IndexMap, IndexSet};
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{safe_assert, safe_unwrap, safe_unwrap_err};
 use move_core_types::{
     account_address::AccountAddress,
     annotated_value::{MoveFieldLayout, MoveStructLayout, MoveTypeLayout, MoveValue},
@@ -100,8 +101,8 @@ pub fn end_transaction(
     ty_args: Vec<Type>,
     args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
-    assert!(args.is_empty());
+    safe_assert!(ty_args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime_ref: &mut ObjectRuntime = get_extension_mut!(context)?;
     let taken_shared_or_imm: BTreeMap<_, _> = object_runtime_ref
         .test_inventories
@@ -280,14 +281,14 @@ pub fn end_transaction(
         new_object_values
             .iter()
             .map(|(id, (ty, value))| (id, ty, value)),
-    );
+    )?;
     find_all_wrapped_objects(
         context,
         &mut all_wrapped,
         object_runtime_ref
             .all_active_child_objects()
             .filter_map(|child| Some((child.id, child.ty, child.copied_value?))),
-    );
+    )?;
     // mark as "incorrect" if a shared/imm object was wrapped or is a child object
     incorrect_shared_or_imm_handling = incorrect_shared_or_imm_handling
         || taken_shared_or_imm.keys().any(|id| {
@@ -368,7 +369,7 @@ pub fn end_transaction(
         transferred,
         user_events.len() as u64,
         // TODO: do we need accumulator events here?
-    );
+    )?;
     Ok(NativeResult::ok(legacy_test_cost(), smallvec![effects]))
 }
 
@@ -378,11 +379,11 @@ pub fn take_from_address_by_id(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
+    let specified_ty = get_specified_ty(ty_args)?;
     let id = pop_id(&mut args)?;
     let account: SuiAddress = pop_arg!(args, AccountAddress).into();
     pop_arg!(args, StructRef);
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -413,9 +414,9 @@ pub fn ids_for_address(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
+    let specified_ty = get_specified_ty(ty_args)?;
     let account: SuiAddress = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -425,7 +426,7 @@ pub fn ids_for_address(
         .and_then(|inv| inv.get(&specified_obj_ty))
         .map(|s| s.iter().map(|id| pack_id(*id)).collect::<Vec<Value>>())
         .unwrap_or_default();
-    let ids_vector = Vector::pack(VectorSpecialization::Container, ids).unwrap();
+    let ids_vector = safe_unwrap_err!(Vector::pack(VectorSpecialization::Container, ids));
     Ok(NativeResult::ok(legacy_test_cost(), smallvec![ids_vector]))
 }
 
@@ -435,15 +436,15 @@ pub fn most_recent_id_for_address(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
+    let specified_ty = get_specified_ty(ty_args)?;
     let account: SuiAddress = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
     let most_recent_id = match inventories.address_inventories.get(&account) {
-        None => pack_option(vector_specialization(&specified_ty), None),
-        Some(inv) => most_recent_at_ty(&inventories.taken, inv, &specified_ty, specified_obj_ty),
+        None => pack_option(vector_specialization(&specified_ty), None)?,
+        Some(inv) => most_recent_at_ty(&inventories.taken, inv, &specified_ty, specified_obj_ty)?,
     };
     Ok(NativeResult::ok(
         legacy_test_cost(),
@@ -457,10 +458,10 @@ pub fn was_taken_from_address(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
+    safe_assert!(ty_args.is_empty());
     let id = pop_id(&mut args)?;
     let account: SuiAddress = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
     let was_taken = inventories
@@ -480,10 +481,10 @@ pub fn take_immutable_by_id(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
+    let specified_ty = get_specified_ty(ty_args)?;
     let id = pop_id(&mut args)?;
     pop_arg!(args, StructRef);
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -520,8 +521,8 @@ pub fn most_recent_immutable_id(
     ty_args: Vec<Type>,
     args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
-    assert!(args.is_empty());
+    let specified_ty = get_specified_ty(ty_args)?;
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -530,7 +531,7 @@ pub fn most_recent_immutable_id(
         &inventories.immutable_inventory,
         &specified_ty,
         specified_obj_ty,
-    );
+    )?;
     Ok(NativeResult::ok(
         legacy_test_cost(),
         smallvec![most_recent_id],
@@ -543,9 +544,9 @@ pub fn was_taken_immutable(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
+    safe_assert!(ty_args.is_empty());
     let id = pop_id(&mut args)?;
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
     let was_taken = inventories
@@ -565,10 +566,10 @@ pub fn take_shared_by_id(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
+    let specified_ty = get_specified_ty(ty_args)?;
     let id = pop_id(&mut args)?;
     pop_arg!(args, StructRef);
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -598,8 +599,8 @@ pub fn most_recent_id_shared(
     ty_args: Vec<Type>,
     args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let specified_ty = get_specified_ty(ty_args);
-    assert!(args.is_empty());
+    let specified_ty = get_specified_ty(ty_args)?;
+    safe_assert!(args.is_empty());
     let specified_obj_ty = object_type_of_type(context, &specified_ty)?;
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
@@ -608,7 +609,7 @@ pub fn most_recent_id_shared(
         &inventories.shared_inventory,
         &specified_ty,
         specified_obj_ty,
-    );
+    )?;
     Ok(NativeResult::ok(
         legacy_test_cost(),
         smallvec![most_recent_id],
@@ -621,9 +622,9 @@ pub fn was_taken_shared(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    assert!(ty_args.is_empty());
+    safe_assert!(ty_args.is_empty());
     let id = pop_id(&mut args)?;
-    assert!(args.is_empty());
+    safe_assert!(args.is_empty());
     let object_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     let inventories = &mut object_runtime.test_inventories;
     let was_taken = inventories
@@ -642,7 +643,7 @@ pub fn allocate_receiving_ticket_for_object(
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    let ty = get_specified_ty(ty_args);
+    let ty = get_specified_ty(ty_args)?;
     let id = pop_id(&mut args)?;
 
     let abilities = context.type_to_abilities(&ty)?;
@@ -662,7 +663,7 @@ pub fn allocate_receiving_ticket_for_object(
         ));
     }
 
-    let obj_value = inventories.objects.remove(&id).unwrap();
+    let obj_value = safe_unwrap!(inventories.objects.remove(&id));
     let Some(bytes) = obj_value.typed_serialize(&layout) else {
         return Ok(NativeResult::err(
             context.gas_used(),
@@ -670,7 +671,7 @@ pub fn allocate_receiving_ticket_for_object(
         ));
     };
     let has_public_transfer = abilities.has_store();
-    let move_object = unsafe {
+    let move_object = safe_unwrap_err!(unsafe {
         MoveObject::new_from_execution_with_limit(
             tag.into(),
             has_public_transfer,
@@ -678,8 +679,7 @@ pub fn allocate_receiving_ticket_for_object(
             bytes,
             250 * 1024,
         )
-    }
-    .unwrap();
+    });
 
     let Some((owner, _)) = inventories
         .address_inventories
@@ -765,17 +765,18 @@ fn take_from_inventory(
     id: ObjectID,
     owner: Owner,
 ) -> Result<Value, NativeResult> {
-    let obj_opt = objects.get(&id);
     let is_taken = taken.contains_key(&id);
-    if is_taken || !is_in_inventory(&id) || obj_opt.is_none() {
-        return Err(NativeResult::err(
-            legacy_test_cost(),
-            E_OBJECT_NOT_FOUND_CODE,
-        ));
-    }
+    let obj = match objects.get(&id) {
+        Some(obj) if !is_taken && is_in_inventory(&id) => obj,
+        _ => {
+            return Err(NativeResult::err(
+                legacy_test_cost(),
+                E_OBJECT_NOT_FOUND_CODE,
+            ));
+        }
+    };
     taken.insert(id, owner.clone());
     input_objects.insert(id, owner);
-    let obj = obj_opt.unwrap();
     Ok(obj.copy_value())
 }
 
@@ -794,7 +795,7 @@ fn most_recent_at_ty(
     inv: &BTreeMap<MoveObjectType, Set<ObjectID>>,
     runtime_ty: &Type,
     ty: MoveObjectType,
-) -> Value {
+) -> PartialVMResult<Value> {
     pack_option(
         vector_specialization(runtime_ty),
         most_recent_at_ty_opt(taken, inv, ty),
@@ -811,9 +812,9 @@ fn most_recent_at_ty_opt(
     Some(pack_id(*most_recent_id))
 }
 
-fn get_specified_ty(mut ty_args: Vec<Type>) -> Type {
-    assert!(ty_args.len() == 1);
-    ty_args.pop().unwrap()
+fn get_specified_ty(mut ty_args: Vec<Type>) -> PartialVMResult<Type> {
+    safe_assert!(ty_args.len() == 1);
+    Ok(safe_unwrap!(ty_args.pop()))
 }
 
 // helpers
@@ -835,24 +836,20 @@ fn pack_id(a: impl Into<AccountAddress>) -> Value {
     Value::struct_(values::Struct::pack(vec![Value::address(a.into())]))
 }
 
-fn pack_ids(items: impl IntoIterator<Item = impl Into<AccountAddress>>) -> Value {
+fn pack_ids(items: impl IntoIterator<Item = impl Into<AccountAddress>>) -> PartialVMResult<Value> {
     Vector::pack(
         VectorSpecialization::Container,
         items.into_iter().map(pack_id),
     )
-    .unwrap()
 }
 
-fn pack_vec_map(items: impl IntoIterator<Item = (Value, Value)>) -> Value {
-    Value::struct_(values::Struct::pack(vec![
-        Vector::pack(
-            VectorSpecialization::Container,
-            items
-                .into_iter()
-                .map(|(k, v)| Value::struct_(values::Struct::pack(vec![k, v]))),
-        )
-        .unwrap(),
-    ]))
+fn pack_vec_map(items: impl IntoIterator<Item = (Value, Value)>) -> PartialVMResult<Value> {
+    Ok(Value::struct_(values::Struct::pack(vec![Vector::pack(
+        VectorSpecialization::Container,
+        items
+            .into_iter()
+            .map(|(k, v)| Value::struct_(values::Struct::pack(vec![k, v]))),
+    )?])))
 }
 
 fn transaction_effects(
@@ -861,7 +858,7 @@ fn transaction_effects(
     deleted: impl IntoIterator<Item = impl Into<AccountAddress>>,
     transferred: impl IntoIterator<Item = (ObjectID, Owner)>,
     num_events: u64,
-) -> Value {
+) -> PartialVMResult<Value> {
     let mut transferred_to_account = vec![];
     let mut transferred_to_object = vec![];
     let mut shared = vec![];
@@ -880,15 +877,15 @@ fn transaction_effects(
         }
     }
 
-    let created_field = pack_ids(created);
-    let written_field = pack_ids(written);
-    let deleted_field = pack_ids(deleted);
-    let transferred_to_account_field = pack_vec_map(transferred_to_account);
-    let transferred_to_object_field = pack_vec_map(transferred_to_object);
-    let shared_field = pack_ids(shared);
-    let frozen_field = pack_ids(frozen);
+    let created_field = pack_ids(created)?;
+    let written_field = pack_ids(written)?;
+    let deleted_field = pack_ids(deleted)?;
+    let transferred_to_account_field = pack_vec_map(transferred_to_account)?;
+    let transferred_to_object_field = pack_vec_map(transferred_to_object)?;
+    let shared_field = pack_ids(shared)?;
+    let frozen_field = pack_ids(frozen)?;
     let num_events_field = Value::u64(num_events);
-    Value::struct_(values::Struct::pack(vec![
+    Ok(Value::struct_(values::Struct::pack(vec![
         created_field,
         written_field,
         deleted_field,
@@ -897,7 +894,7 @@ fn transaction_effects(
         shared_field,
         frozen_field,
         num_events_field,
-    ]))
+    ])))
 }
 
 fn object_type_of_type(context: &NativeContext, ty: &Type) -> PartialVMResult<MoveObjectType> {
@@ -909,21 +906,22 @@ fn object_type_of_type(context: &NativeContext, ty: &Type) -> PartialVMResult<Mo
     Ok(MoveObjectType::from(*s_tag))
 }
 
-fn pack_option(specialization: VectorSpecialization, opt: Option<Value>) -> Value {
+fn pack_option(specialization: VectorSpecialization, opt: Option<Value>) -> PartialVMResult<Value> {
     let item = match opt {
         Some(v) => vec![v],
         None => vec![],
     };
-    Value::struct_(values::Struct::pack(vec![
-        Vector::pack(specialization, item).unwrap(),
-    ]))
+    Ok(Value::struct_(values::Struct::pack(vec![Vector::pack(
+        specialization,
+        item,
+    )?])))
 }
 
 fn find_all_wrapped_objects<'a, 'i>(
     context: &NativeContext,
     ids: &'i mut BTreeSet<ObjectID>,
     new_object_values: impl IntoIterator<Item = (&'a ObjectID, &'a MoveObjectType, impl Borrow<Value>)>,
-) {
+) -> PartialVMResult<()> {
     #[derive(Copy, Clone)]
     enum LookingFor {
         Wrapped,
@@ -1013,8 +1011,8 @@ fn find_all_wrapped_objects<'a, 'i>(
             continue;
         };
 
-        let blob = value.borrow().typed_serialize(&layout).unwrap();
-        MoveValue::visit_deserialize(
+        let blob = safe_unwrap!(value.borrow().typed_serialize(&layout));
+        safe_unwrap_err!(MoveValue::visit_deserialize(
             &blob,
             &annotated_layout,
             &mut Traversal {
@@ -1022,7 +1020,7 @@ fn find_all_wrapped_objects<'a, 'i>(
                 ids,
                 uid: &uid,
             },
-        )
-        .unwrap();
+        ));
     }
+    Ok(())
 }
diff --git a/sui-execution/latest/sui-move-natives/src/test_utils.rs b/sui-execution/latest/sui-move-natives/src/test_utils.rs
index d21862f34bc2f..90663e485f64c 100644
--- a/sui-execution/latest/sui-move-natives/src/test_utils.rs
+++ b/sui-execution/latest/sui-move-natives/src/test_utils.rs
@@ -3,6 +3,7 @@
 
 use crate::{legacy_test_cost, types::is_otw_struct};
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::safe_unwrap;
 use move_core_types::{gas_algebra::InternalGas, runtime_value::MoveTypeLayout};
 use move_vm_runtime::execution::values::Struct;
 use move_vm_runtime::execution::{Type, values::Value};
@@ -18,7 +19,7 @@ pub fn create_one_time_witness(
     debug_assert!(ty_args.len() == 1);
     debug_assert!(args.is_empty());
 
-    let ty = ty_args.pop().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
     let type_tag = context.type_to_type_tag(&ty)?;
     let type_layout = context.type_to_type_layout(&ty)?;
 
diff --git a/sui-execution/latest/sui-move-natives/src/transfer.rs b/sui-execution/latest/sui-move-natives/src/transfer.rs
index 6c381e8903ada..32cc718257564 100644
--- a/sui-execution/latest/sui-move-natives/src/transfer.rs
+++ b/sui-execution/latest/sui-move-natives/src/transfer.rs
@@ -7,6 +7,7 @@ use crate::{
     get_tag_and_layouts, object_runtime::object_store::ObjectResult,
 };
 use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_binary_format::{safe_assert, safe_unwrap, safe_unwrap_err};
 use move_core_types::{
     account_address::AccountAddress, gas_algebra::InternalGas, language_storage::TypeTag,
     vm_status::StatusCode,
@@ -58,7 +59,7 @@ pub fn receive_object_internal(
         context,
         transfer_receive_object_internal_cost_params.transfer_receive_object_internal_cost_base
     );
-    let child_ty = ty_args.pop().unwrap();
+    let child_ty = safe_unwrap!(ty_args.pop());
     native_charge_gas_early_exit!(
         context,
         transfer_receive_object_internal_cost_params
@@ -66,15 +67,15 @@ pub fn receive_object_internal(
             * u64::from(child_ty.size()?).into()
     );
     let child_receiver_sequence_number: SequenceNumber = pop_arg!(args, u64).into();
-    let child_receiver_object_id = args.pop_back().unwrap();
+    let child_receiver_object_id = safe_unwrap!(args.pop_back());
     let parent = pop_arg!(args, AccountAddress).into();
-    assert!(args.is_empty());
-    let child_id: ObjectID = get_receiver_object_id(child_receiver_object_id.copy_value())
-        .unwrap()
-        .value_as::<AccountAddress>()
-        .unwrap()
-        .into();
-    assert!(ty_args.is_empty());
+    safe_assert!(args.is_empty());
+    let child_id: ObjectID = safe_unwrap_err!(
+        get_receiver_object_id(child_receiver_object_id.copy_value())
+            .and_then(|v| v.value_as::<AccountAddress>())
+    )
+    .into();
+    safe_assert!(ty_args.is_empty());
 
     let Some((tag, layout, annotated_layout)) = get_tag_and_layouts(context, &child_ty)? else {
         return Ok(NativeResult::err(
@@ -149,9 +150,9 @@ pub fn transfer_internal(
         transfer_transfer_internal_cost_params.transfer_transfer_internal_cost_base
     );
 
-    let ty = ty_args.pop().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
     let recipient = pop_arg!(args, AccountAddress);
-    let obj = args.pop_back().unwrap();
+    let obj = safe_unwrap!(args.pop_back());
 
     let owner = Owner::AddressOwner(recipient.into());
     object_runtime_transfer(context, owner, ty, obj)?;
@@ -215,11 +216,11 @@ pub fn party_transfer_internal(
         transfer_party_transfer_internal_cost_params.transfer_party_transfer_internal_cost_base
     );
 
-    let ty = ty_args.pop().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
     let permissions = pop_arg!(args, Vec<u64>);
     let addresses = pop_arg!(args, Vec<AccountAddress>);
     let default_permissions = pop_arg!(args, u64);
-    let obj = args.pop_back().unwrap();
+    let obj = safe_unwrap!(args.pop_back());
     let Ok([permissions]): Result<[u64; 1], _> = permissions.try_into() else {
         return Err(
             PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
@@ -277,8 +278,8 @@ pub fn freeze_object(
         transfer_freeze_object_cost_params.transfer_freeze_object_cost_base
     );
 
-    let ty = ty_args.pop().unwrap();
-    let obj = args.pop_back().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
+    let obj = safe_unwrap!(args.pop_back());
 
     object_runtime_transfer(context, Owner::Immutable, ty, obj)?;
 
@@ -311,8 +312,8 @@ pub fn share_object(
         transfer_share_object_cost_params.transfer_share_object_cost_base
     );
 
-    let ty = ty_args.pop().unwrap();
-    let obj = args.pop_back().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
+    let obj = safe_unwrap!(args.pop_back());
     let transfer_result = object_runtime_transfer(
         context,
         // Dummy version, to be filled with the correct initial version when the effects of the
diff --git a/sui-execution/latest/sui-move-natives/src/tx_context.rs b/sui-execution/latest/sui-move-natives/src/tx_context.rs
index 4e0a9df85643f..5b0855b7216b2 100644
--- a/sui-execution/latest/sui-move-natives/src/tx_context.rs
+++ b/sui-execution/latest/sui-move-natives/src/tx_context.rs
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::safe_unwrap_err;
 use move_core_types::{account_address::AccountAddress, gas_algebra::InternalGas};
 use move_vm_runtime::{
     execution::{Type, values::Value},
@@ -46,8 +47,7 @@ pub fn derive_id(
     let ids_created = pop_arg!(args, u64);
     let tx_hash = pop_arg!(args, Vec<u8>);
 
-    // unwrap safe because all digests in Move are serialized from the Rust `TransactionDigest`
-    let digest = TransactionDigest::try_from(tx_hash.as_slice()).unwrap();
+    let digest = safe_unwrap_err!(TransactionDigest::try_from(tx_hash.as_slice()));
     let address = AccountAddress::from(ObjectID::derive_id(digest, ids_created));
     let obj_runtime: &mut ObjectRuntime = get_extension_mut!(context)?;
     obj_runtime.new_id(address.into())?;
diff --git a/sui-execution/latest/sui-move-natives/src/types.rs b/sui-execution/latest/sui-move-natives/src/types.rs
index 34003a78a841a..34da32a54f9c1 100644
--- a/sui-execution/latest/sui-move-natives/src/types.rs
+++ b/sui-execution/latest/sui-move-natives/src/types.rs
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use move_binary_format::errors::PartialVMResult;
+use move_binary_format::safe_unwrap;
 use move_core_types::{
     gas_algebra::InternalGas,
     language_storage::TypeTag,
@@ -68,8 +69,7 @@ pub fn is_one_time_witness(
         type_is_one_time_witness_cost_params.types_is_one_time_witness_cost_base
     );
 
-    // unwrap safe because the interface of native function guarantees it.
-    let ty = ty_args.pop().unwrap();
+    let ty = safe_unwrap!(ty_args.pop());
 
     native_charge_gas_early_exit!(
         context,
