diff --git a/README.md b/README.md
index a6d52794..7e41fe8e 100644
--- a/README.md
+++ b/README.md
@@ -130,7 +130,7 @@ CUDA_VISIBLE_DEVICES=1 uv run train @ configs/training/simple_math.toml
 
 *TBD*
 
-## Contributing
+## Developer
 
 *For now, development is only possible on CUDA-enabled devices. However, we build production-ready images for both CUDA (NVIDIA) and ROCM (AMD) GPUs that should work out of the box.*
 
@@ -142,7 +142,7 @@ CUDA_VISIBLE_DEVICES=1 uv run train @ configs/training/simple_math.toml
 uv run pre-commit install
 ```
 
-### Config System
+### Configs
 
 We use `pydantic-settings` to configure `prime-rl`. To get an overview of the available configurations, run the following command:
 
@@ -188,6 +188,59 @@ PRIME_MODEL__NAME=Qwen/Qwen3-4B uv run src/zeroband/inference/server.py @qwen8b.
 
 In this example, the CLI argument `--model.name Qwen/Qwen3-32B` will take precendence and the script will use `Qwen/Qwen3-32B` as the model name. If the CLI argument wasn't set, then the second config file would take precedence and the script would use `Qwen/Qwen-14B` as the model name. If the second config file wasn't set, then the first config file would take precedence and the script would use `Qwen/Qwen3-8B` as the model name. Finally, if the first config file wasn't set, then the environment variable would take precedence and the script would use `Qwen/Qwen-4B` as the model name. If the environment variable wasn't set, then the default value would be used and the script would use `Qwen/Qwen3-0.6B` as the model name.
 
+### Checkpointing
+
+Our codebase supports checkpointing. Because of the trainer/ orchestrator design, as well as the natural asynchrony checkpointing is non-standard.
+
+- Trainer (`src/zeroband/training/ckpt.py`): Checkpoints FSDP model shard, optimizer state and progress (training step, total samples, total tokens)
+- Orchestrator (`src/zeroband/training/ckpt.py`): Checkpoints orchestrator progress
+
+*NB: Each run with asynchrony level `async_level` and some checkpoint step `x`, requires weight checkpoints in the step range `[x-async_level, x]`. Currently we do not duplicate weight checkpoints into the `checkpoints` directory but simply keep them around in `weights`, by keeping the trainer from cleaning up weight checkpoints that are required for resuming training. This way, the orchestrator only needs to checkpoint its progress (read: step) to load the correct weights into the inference engine upon resuming.*
+
+The default checkpoint directory is `checkpoints` and each checkpoint step will live in a subdirectory enumerated by the step, i.e. `checkpoints/step_{step}`. The trainer checkpoint is called `trainer.pt` for single GPU workloads, else `trainer_{local_rank}.pt`. The orchestrator checkpoint is callec `orchestrator.pt`. Thus, this is a typical directory structure:
+
+```bash
+checkpoints
+├── step_10
+│   ├── orchestrator.pt
+│   └── trainer.pt
+├── step_25
+│   ├── orchestrator.pt
+│   └── trainer.pt
+└── step_30
+    ├── orchestrator.pt
+    └── trainer.pt
+```
+
+Checkpointing is configured by the `CheckpointConfig`, with the config key `--ckpt`. One can specify:
+- `--ckpt.path` to change the checkpoint directory (default: `checkpoints`)
+- `--ckpt.interval` to change the interval frequency (default: `50`)
+- `--ckpt.save-async` to save the checkpoint asynchronously (default: `False`)
+
+By default, runs do no write checkpoints to save disk space. To checkpoint every 10 steps on our debug RL run, run the following command
+
+```bash
+CUDA_VISIBLE_DEVICES=1 uv run train @ configs/training/reverse_text.toml --ckpt.interval 10 
+```
+
+To resume a run use the `--ckpt.resume-step` flag. To resume from the checkpoint stpe 10 from the previous command, run the following command
+
+```bash
+CUDA_VISIBLE_DEVICES=1 uv run train @ configs/training/reverse_text.toml --ckpt.resume_step 10
+```
+
+Because we save progress information, resuming from a checkpoint is fully W&B compatible. By default, resuming from a checkpoint, will simply create a new run. To resume the same W&B run, you'd have to pass the same W&B run ID for both the trainer and the orchestrator, e.g.
+
+```bash
+CUDA_VISIBLE_DEVICES=1 uv run train @ configs/training/reverse_text.toml \
+  --monitor.wandb.project <project> \
+  --monitor.wandb.group <group> \
+  --ckpt.resume-step 10 \
+  --monitor.wandb.id <trainer-run-id> \
+  --orchestrator.monitor.wandb.id <orchestrator-run-id>
+```
+
+
 ### Tests
 
 Run the full test suite 
diff --git a/src/zeroband/training/ckpt.py b/src/zeroband/training/ckpt.py
index 6fba1d2f..a189029c 100644
--- a/src/zeroband/training/ckpt.py
+++ b/src/zeroband/training/ckpt.py
@@ -1,150 +1,106 @@
 import threading
 import time
-from dataclasses import dataclass
+from copy import deepcopy
+from dataclasses import asdict, dataclass
 from pathlib import Path
 
 import torch
-from torch.distributed.checkpoint.state_dict import _get_fqns as get_fqns
-from torch.distributed.tensor import DTensor
 from torch.optim.optimizer import Optimizer
-from transformers import AutoTokenizer
 
+from zeroband.training.config import CheckpointConfig
 from zeroband.training.model import Model
 from zeroband.training.world import get_world
 from zeroband.utils.logger import get_logger
 
 
 @dataclass
-class TrainingProgress:
+class Progress:
     step: int = 0
     total_tokens: int = 0
     total_samples: int = 0
 
 
-def save_full_checkpoint(
-    model: Model,
-    optimizers: list[Optimizer],
-    progress: TrainingProgress,
-    path: Path,
-):
-    # Get logger
-    logger = get_logger()
-    start_time = time.time()
-    logger.debug(f"Writing checkpoint to {path}")
-
-    # Create checkpoint state
-    ckpt_state = {
-        "model": model.state_dict(),
-        "optimizers": [optimizer.state_dict() for optimizer in optimizers],
-        "progress": progress,
-    }
-
-    # Create checkpoint directory if it doesn't exist
-    path.mkdir(parents=True, exist_ok=True)
-    local_path = path / f"local_rank_{get_world().local_rank}"
-    with open(local_path, "wb") as f:
-        torch.save(ckpt_state, f)
-    logger.debug(f"Checkpoint saved at {path} in {time.time() - start_time:.2f} seconds")
-
-
-def load_full_checkpoint(
-    model: Model,
-    optimizers: list[Optimizer],
-    progress: TrainingProgress,
-    path: Path,
-):
-    # Get logger
-    logger = get_logger()
-    start_time = time.time()
-    logger.debug(f"Loading checkpoint from {path}")
-
-    # Check local step path exists
-    local_path = path / f"local_rank_{get_world().local_rank}"
-    if not local_path.exists():
-        raise FileNotFoundError(f"Checkpoint step {progress.step} not found at {local_path}")
-
-    # Load checkpoint state
-    with open(local_path, "rb") as f:
-        state = torch.load(f, weights_only=False)
-
-    # Initialize model and optimizers
-    model.load_state_dict(state["model"])
-    for optimizer, optimizer_state in zip(optimizers, state["optimizers"]):
-        optimizer.load_state_dict(optimizer_state)
-
-    # Update progress
-    progress.total_tokens = state["progress"].total_tokens
-    progress.step = state["progress"].step
-    progress.total_samples = state["progress"].total_samples
-
-    logger.debug(f"Checkpoint loaded in {time.time() - start_time:.2f} seconds")
-
-
-def save_weight_checkpoint(
-    model: Model,
-    tokenizer: AutoTokenizer,
-    path: Path,
-    dtype: torch.dtype = torch.bfloat16,
-    async_save: bool = False,
-) -> Path:
-    """
-    Save a HF-compatible weight-only checkpoint to the specified path. Saves the
-    model weights as `model.pt`, the model config, generation arguments and tokenizer
-    using HF's `save_pretrained` method.
-    """
-    # Get logger and world info
-    logger = get_logger()
-    is_master = get_world().rank == 0
-
-    # Create checkpoint directory if it doesn't exist
-    if not path.exists():
-        path.mkdir(parents=True, exist_ok=True)
-
-    # Gather distributed weights for weight checkpoint
-    start_time = time.time()
-    logger.debug("Gathering sharded weights")
-    cpu_state = {}
-    for key, value in model.state_dict().items():
-        if isinstance(value, DTensor):
-            value = value.to(dtype)
-            # only gather after the downcast to dtype as it will be faster
-            value = value.full_tensor()
-
-        if is_master:
-            key = get_fqns(model, key)
-            assert len(key) == 1
-            key = next(iter(key))
-            # TODO(Sami) Blocking to avoid race condition, should make non-blocking long-term tho
-            cpu_state[key] = value.to("cpu", non_blocking=False)
-
-    torch.distributed.barrier()
-    logger.debug(f"Gathered sharded weights in {time.time() - start_time:.2f} seconds")
-
-    model_path = path / "model.pt"
-
-    def _save_weight_checkpoint():
-        if is_master:
-            start_time = time.time()
-            logger.debug(f"Saving weights to {path}")
-
-            # Save model weights to temporary file to avoid race condition
-            tmp_model_path = path / "model.pt.tmp"
-            torch.save(cpu_state, tmp_model_path)
-
-            # Rename temporary file to indicate checkpoint is complete
-            tmp_model_path.rename(model_path)
-
-            # Save model config, generation arguments and tokenizer
-            model.config.save_pretrained(path)
-            model.generation_config.save_pretrained(path)
-            tokenizer.save_pretrained(path)
-
-            logger.debug(f"Saved weights to {path} in {time.time() - start_time:.2f} seconds")
-
-    if async_save:
-        thread = threading.Thread(target=_save_weight_checkpoint)
-        thread.start()
-    else:
-        _save_weight_checkpoint()
-
-    return model_path
+class CheckpointManager:
+    """Utility class to save and load training checkpoints to resume training."""
+
+    def __init__(self, config: CheckpointConfig):
+        self.path = config.path
+        self.save_async = config.save_async
+        self._logger = get_logger()
+        self._world = get_world()
+
+    def _get_step_path(self, step: int) -> Path:
+        return self.path / f"step_{step}"
+
+    def _get_ckpt_path(self, step: int) -> Path:
+        ckpt_name = f"trainer_{self._world.local_rank}.pt" if self._world.world_size > 1 else "trainer.pt"
+        return self._get_step_path(step) / ckpt_name
+
+    def _save_to_path(self, ckpt_path: Path, model: Model, optimizers: list[Optimizer], progress: Progress):
+        self._logger.debug(f"Saving training checkpoint to {ckpt_path}")
+        start_time = time.time()
+
+        # Increment the progress step that is going to be saved
+        progress_copy = deepcopy(progress)
+        progress_copy.step += 1
+
+        # Create checkpoint state
+        ckpt_state = {
+            "model": model.state_dict(),
+            "optimizers": [optimizer.state_dict() for optimizer in optimizers],
+            "progress": progress_copy,
+        }
+        # Create checkpoint directory if it doesn't exist
+        ckpt_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(ckpt_path, "wb") as f:
+            torch.save(ckpt_state, f)
+        self._logger.debug(f"Training checkpoint saved in {time.time() - start_time:.2f} seconds")
+
+    def _load_from_path(self, ckpt_path: Path, model: Model, optimizers: list[Optimizer], progress: Progress):
+        """Loads a checkpoint from a given path in-place."""
+        self._logger.debug(f"Loading training checkpoint from {ckpt_path}")
+        start_time = time.time()
+
+        # Load checkpoint state
+        with open(ckpt_path, "rb") as f:
+            state = torch.load(f, weights_only=False)
+
+        # Load checkpoint state in-place
+        model.load_state_dict(state["model"])
+        for optimizer, optimizer_state in zip(optimizers, state["optimizers"]):
+            optimizer.load_state_dict(optimizer_state)
+
+        # Load progress
+        for key, value in asdict(state["progress"]).items():
+            setattr(progress, key, value)
+
+        self._logger.debug(f"Training checkpoint loaded in {time.time() - start_time:.2f} seconds")
+
+    def load(self, model: Model, optimizers: list[Optimizer], progress: Progress, step: int) -> None:
+        """Loads a checkpoint from a given path in-place."""
+        ckpt_path = self._get_ckpt_path(step)
+        if not ckpt_path.exists():
+            raise FileNotFoundError(f"Checkpoint not found at {ckpt_path}")
+        self._load_from_path(ckpt_path, model, optimizers, progress)
+
+    def save(
+        self,
+        model: Model,
+        optimizers: list[Optimizer],
+        progress: dict,
+        step: int,
+    ):
+        """Saves the full checkpoint state for a specified step."""
+        step_path = self._get_step_path(step)
+        step_path.mkdir(parents=True, exist_ok=True)
+        ckpt_path = self._get_ckpt_path(step)
+
+        if self.save_async:
+            # Run save in a separate thread
+            thread = threading.Thread(
+                target=self._save_to_path, args=(ckpt_path, model, optimizers, progress), name=f"ckpt-save-{step}"
+            )
+            thread.start()
+        else:
+            # Run save synchronously
+            self._save_to_path(ckpt_path, model, optimizers, progress)
diff --git a/src/zeroband/training/config.py b/src/zeroband/training/config.py
index 9b33947d..49b411b7 100644
--- a/src/zeroband/training/config.py
+++ b/src/zeroband/training/config.py
@@ -1,3 +1,4 @@
+import warnings
 from pathlib import Path
 from typing import Annotated, Literal, TypeAlias, Union
 
@@ -53,21 +54,22 @@ class OptimizerConfig(BaseConfig):
 class CheckpointConfig(BaseConfig):
     """Configures checkpointing the full model, optimizer and training state for resuming training."""
 
-    path: Annotated[Path, Field(description="Path to write checkpoints to.")] = Path("checkpoints")
+    path: Annotated[Path, Field(description="Directory to write checkpoints to.")] = Path("checkpoints")
 
-    clean: Annotated[
+    interval: Annotated[int, Field(ge=1, description="Interval at which to save the checkpoint.")] = 50
+
+    save_async: Annotated[
         bool,
         Field(
-            description="Whether to clean the path at the beginning of the run. If True, will delete the entire directory.",
+            description="Whether to save the checkpoint asynchronously.",
         ),
     ] = False
 
-    interval: Annotated[int, Field(ge=1, description="Interval at which to save the checkpoint.")] = 50
-
-    resume_path: Annotated[
-        Path | None,
+    resume_step: Annotated[
+        int | None,
         Field(
-            description="Checkpoint path to resume training from. If None, will start from scratch.",
+            ge=1,
+            description="Step to resume training from. If None, will start from scratch.",
         ),
     ] = None
 
@@ -92,9 +94,9 @@ class WeightCheckpointConfig(BaseConfig):
     save_async: Annotated[
         bool,
         Field(
-            description="Whether to save the checkpoint asynchronously.",
+            description="Whether to save the weights asynchronously.",
         ),
-    ] = True
+    ] = False
 
 
 class BaseGRPOVariantConfig(BaseConfig):
@@ -174,7 +176,7 @@ class TrainingConfig(BaseSettings):
     optim: OptimizerConfig = OptimizerConfig()
 
     # The checkpoint configuration
-    ckpt: CheckpointConfig = CheckpointConfig(path=Path("checkpoints"), clean=True)
+    ckpt: CheckpointConfig | None = None
 
     # The weight checkpoint configuration
     weights: WeightCheckpointConfig = WeightCheckpointConfig()
@@ -239,3 +241,34 @@ class TrainingConfig(BaseSettings):
         if self.orchestrator:
             self.orchestrator.log.level = self.log.level
         return self
+
+    @model_validator(mode="after")
+    def auto_setup_max_step(self):
+        if self.max_steps is not None and self.orchestrator:
+            self.orchestrator.max_steps = self.max_steps
+        return self
+
+    @model_validator(mode="after")
+    def auto_setup_orchestrator_ckpt(self):
+        if self.ckpt:
+            # Ensures that trainer and orchestrator checkpoints are synchronized
+            self.orchestrator.ckpt = CheckpointConfig()
+            self.orchestrator.ckpt.path = self.ckpt.path
+            self.orchestrator.ckpt.interval = self.ckpt.interval
+
+            if self.ckpt.resume_step and self.orchestrator.ckpt.resume_step is None:
+                self.orchestrator.ckpt.resume_step = self.ckpt.resume_step
+        return self
+
+    @model_validator(mode="after")
+    def warn_wandb_resume_id_missing(self):
+        if self.ckpt and self.ckpt.resume_step:
+            if self.monitor.wandb and not self.monitor.wandb.id:
+                warnings.warn(
+                    "W&B run ID is not set for trainer even though resuming training. The current run will be created as a new run."
+                )
+            if self.orchestrator.monitor.wandb and not self.orchestrator.monitor.wandb.id:
+                warnings.warn(
+                    "W&B run ID is not set for orchestrator even though resuming training. The current run will be created as a new run."
+                )
+        return self
diff --git a/src/zeroband/training/orchestrator/ckpt.py b/src/zeroband/training/orchestrator/ckpt.py
new file mode 100644
index 00000000..da66c3a0
--- /dev/null
+++ b/src/zeroband/training/orchestrator/ckpt.py
@@ -0,0 +1,82 @@
+import time
+from copy import deepcopy
+from dataclasses import asdict, dataclass
+from pathlib import Path
+
+import torch
+
+from zeroband.training.orchestrator.config import CheckpointConfig
+from zeroband.utils.logger import get_logger
+
+
+@dataclass
+class Progress:
+    step: int = 0
+    epoch: int = 0
+    total_tokens: int = 0
+    total_samples: int = 0
+    total_problems: int = 0
+
+
+class CheckpointManager:
+    """Utility class to save and load orchestrator checkpoints to resume orchestrator."""
+
+    def __init__(self, config: CheckpointConfig):
+        self.path = config.path
+        self._logger = get_logger()
+
+    def _get_step_path(self, step: int) -> Path:
+        return self.path / f"step_{step}"
+
+    def _get_ckpt_path(self, step: int) -> Path:
+        return self._get_step_path(step) / "orchestrator.pt"
+
+    def _save_to_path(self, ckpt_path: Path, progress: Progress):
+        self._logger.debug(f"Saving orchestrator checkpoint to {ckpt_path}")
+        start_time = time.time()
+
+        # Increment the progress step that is going to be saved
+        progress_copy = deepcopy(progress)
+        progress_copy.step += 1
+
+        # Create checkpoint state
+        ckpt_state = {"progress": progress_copy}
+
+        # Save checkpoint state
+        with open(ckpt_path, "wb") as f:
+            torch.save(ckpt_state, f)
+
+        self._logger.debug(f"Orchestrator checkpoint saved in {time.time() - start_time:.2f} seconds")
+
+    def _load_from_path(self, ckpt_path: Path, progress: Progress) -> None:
+        """Loads a checkpoint from a given path in-place."""
+        self._logger.debug(f"Loading checkpoint from {ckpt_path}")
+        start_time = time.time()
+
+        # Load checkpoint state
+        with open(ckpt_path, "rb") as f:
+            state = torch.load(f, weights_only=False)
+
+        # Load checkpoint state in-place
+        for key, value in asdict(state["progress"]).items():
+            setattr(progress, key, value)
+
+        self._logger.debug(f"Orchestrator checkpoint loaded in {time.time() - start_time:.2f} seconds")
+
+    def load(self, progress: Progress, step: int) -> Path:
+        """Loads a checkpoint from a given path."""
+        ckpt_path = self._get_ckpt_path(step)
+        if not ckpt_path.exists():
+            raise FileNotFoundError(f"Checkpoint not found at {ckpt_path}")
+        self._load_from_path(ckpt_path, progress)
+
+    def save(
+        self,
+        progress: Progress,
+        step: int,
+    ):
+        """Saves the full checkpoint state for a specified step."""
+        step_path = self._get_step_path(step)
+        step_path.mkdir(parents=True, exist_ok=True)
+        ckpt_path = self._get_ckpt_path(step)
+        self._save_to_path(ckpt_path, progress)
diff --git a/src/zeroband/training/orchestrator/client.py b/src/zeroband/training/orchestrator/client.py
index 1bca7795..88820ce0 100644
--- a/src/zeroband/training/orchestrator/client.py
+++ b/src/zeroband/training/orchestrator/client.py
@@ -61,7 +61,7 @@ async def reload_weights(client: AsyncOpenAI, path: Path, step: int) -> None:
     """Make a HTTP post request to the vLLM server to reload the weights."""
     logger = get_logger()
     url = str(client.base_url)[:-4] + "/reload_weights"
-    model_path = path / f"step_{step}" / "model.pt"
+    model_path = path / f"step_{step}" / "pytorch_model.bin"
     logger.debug(f"Sending request to {url} to reload weights from {model_path}")
     await client.post(url, cast_to=Response, body={"model_path": model_path.as_posix()})
 
diff --git a/src/zeroband/training/orchestrator/config.py b/src/zeroband/training/orchestrator/config.py
index ecfeec51..e4b9a783 100644
--- a/src/zeroband/training/orchestrator/config.py
+++ b/src/zeroband/training/orchestrator/config.py
@@ -149,19 +149,6 @@ class DataConfig(BaseConfig):
     ] = None
 
 
-class PathConfig(BaseConfig):
-    """Configures a path used for input/ output operations"""
-
-    path: Annotated[Path, Field(description="Path to write to.")]
-
-    clean: Annotated[
-        bool,
-        Field(
-            description="Whether to clean the path at the beginning of the run. If True, will delete the entire directory.",
-        ),
-    ] = False
-
-
 class OnlineEvalConfig(BaseConfig):
     """Configures online evaluation."""
 
@@ -201,6 +188,22 @@ class EvalConfig(BaseConfig):
     online: Annotated[OnlineEvalConfig | None, Field(description="Whether to do online evaluation.")] = None
 
 
+class CheckpointConfig(BaseConfig):
+    """Configures checkpointing the orchestrator."""
+
+    path: Annotated[Path, Field(description="Directory to write checkpoints to.")] = Path("checkpoints")
+
+    interval: Annotated[int, Field(ge=1, description="Interval at which to save the checkpoint.")] = 50
+
+    resume_step: Annotated[
+        int | None,
+        Field(
+            ge=1,
+            description="Step to resume orchestrator from. If None, will start from scratch.",
+        ),
+    ] = None
+
+
 class OrchestratorConfig(BaseSettings):
     """Configures the orchestrator for RL training."""
 
@@ -225,6 +228,9 @@ class OrchestratorConfig(BaseSettings):
     # The monitor configuration
     monitor: MultiMonitorConfig = MultiMonitorConfig()
 
+    # The checkpoint configuration
+    ckpt: CheckpointConfig | None = None
+
     collate_mode: Annotated[Literal["packing", "padding"], Field(description="Collate mode to use.")] = "padding"
 
     batch_size: Annotated[int, Field(ge=1, description="Number of samples to train on per step.")] = 128
@@ -242,7 +248,7 @@ class OrchestratorConfig(BaseSettings):
         Field(
             description="Sequence length to use for training. If a sample is shorter than this, it will be padded. If a sequence is longer than this, it will be truncated.",
         ),
-    ] = 1024
+    ] = 2048
 
     # TODO(Mika): This should be automatic from the number of ZMQ connections
     num_train_workers: Annotated[
@@ -265,19 +271,26 @@ class OrchestratorConfig(BaseSettings):
         ),
     ] = 2
 
-    rollout: Annotated[
-        PathConfig,
+    rollout_path: Annotated[
+        Path,
         Field(
             description="Path to write inference outputs to. Will be populated by the orchestrator with responses from inference pool.",
         ),
-    ] = PathConfig(path=Path("rollouts"), clean=True)
+    ] = Path("rollouts")
 
-    weights: Annotated[
-        PathConfig,
+    weights_path: Annotated[
+        Path,
         Field(
             description="Path to read updated model weights from. Will be populated by the trainer.",
         ),
-    ] = PathConfig(path=Path("weights"), clean=True)
+    ] = Path("weights")
+
+    clean: Annotated[
+        bool,
+        Field(
+            description="Whether to clean the rollouts, checkpoint, checkpoint weights and logs directories at the beginning of the run. If True, will forceably, and irreversibly, delete all directories.",
+        ),
+    ] = False
 
     seed: Annotated[int | None, Field(description="Random seed for the orchestrator.")] = None
 
diff --git a/src/zeroband/training/orchestrator/orchestrator.py b/src/zeroband/training/orchestrator/orchestrator.py
index 5839f12f..0bd302e8 100644
--- a/src/zeroband/training/orchestrator/orchestrator.py
+++ b/src/zeroband/training/orchestrator/orchestrator.py
@@ -16,6 +16,7 @@ from datasets import Dataset, load_dataset
 from transformers import AutoTokenizer
 
 from zeroband.eval.utils import run_benchmark
+from zeroband.training.orchestrator.ckpt import CheckpointManager, Progress
 from zeroband.training.orchestrator.client import (
     check_has_model,
     check_health,
@@ -53,13 +54,21 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
     logger.debug(f"EvaluationConfig({config.eval})")
 
     # Prepare paths to communicate with the trainer
-    if config.rollout.clean:
-        logger.debug(f"Cleaning rollout path ({config.rollout.path})")
-        shutil.rmtree(config.rollout.path, ignore_errors=True)
+    if config.clean:
+        logger.info("Cleaning checkpoint, logs, checkpoint weights and rollout directories")
+        if not (config.ckpt and config.ckpt.resume_step): # Only clean if we don't resume
+            logger.debug(f"Cleaning checkpoint path ({config.ckpt.path})")
+            shutil.rmtree(config.ckpt.path, ignore_errors=True)
 
-    if config.weights.clean:
-        logger.debug(f"Cleaning weights path ({config.weights.path})")
-        shutil.rmtree(config.weights.path, ignore_errors=True)
+        logger.debug(f"Cleaning logs path ({config.log.path})")
+        shutil.rmtree(config.log.path, ignore_errors=True)
+
+        if not (config.ckpt and config.ckpt.resume_step): # Only clean if we don't resume
+            logger.debug(f"Cleaning checkpoint weights path ({config.weights_path})")
+            shutil.rmtree(config.weights_path, ignore_errors=True)
+
+        logger.debug(f"Cleaning rollout path ({config.rollout_path})")
+        shutil.rmtree(config.rollout_path, ignore_errors=True)
 
     # Setup client
     logger.info(f"Initializing OpenAI client ({config.client.base_url})")
@@ -79,17 +88,29 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
     await check_has_model(client, config.model.name)
     logger.success("Inference pool ready")
 
-    # Reset weights to base model to allow reusing inference server across runs
-    logger.info("Resetting weights to base model")
-    await reset_weights(client)
+    # Get checkpoint manager
+    if config.ckpt:
+        ckpt_manager = CheckpointManager(config.ckpt)
+
+    # Reset weights to base model if starting from scratch
+    progress = Progress()
+    ckpt_step = 0
+    if config.ckpt and config.ckpt.resume_step:
+        logger.info(f"Resuming training from checkpoint step `{config.ckpt.resume_step}`")
+        ckpt_manager.load(progress, step=config.ckpt.resume_step)
+        ckpt_step = max(progress.step - config.async_level, 0)
+        await reload_weights(client, config.weights_path, ckpt_step)
+    else:
+        logger.info("Training from scratch. Resetting weights to base model")
+        await reset_weights(client)
 
     # Signal that setup is complete
     if setup_queue is not None:
         logger.info("Signaling trainer that orchestrator setup is complete")
         setup_queue.put("ready")
 
-    # Optionally, run evals on base model
-    if config.eval:
+    # Optionally, run evals on base model (only if starting from scratch)
+    if config.eval and not config.ckpt.resume_step:
         logger.info("Running evals on base model")
         for benchmark in config.eval.benchmarks:
             await run_benchmark(client, benchmark, config.model, config.sampling, step=0, use_tqdm=True)
@@ -110,26 +131,22 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
     dataset = dataset.shuffle(seed=config.seed)
 
     # Iterate over dataset in batches
-    max_steps = config.max_steps or int(1e9)
     steps_per_epoch = len(dataset) // (config.batch_size // config.sampling.n)
-    logger.info(f"Starting training loop (max_steps={max_steps}, steps_per_epoch={steps_per_epoch})")
-    total_tokens, total_samples = 0, 0
-    ckpt_step = 0
+    logger.info(f"Starting training loop ({steps_per_epoch=})")
     last_eval_step = -1
-    epoch = -1
+    while True:
+        if config.max_steps and progress.step >= config.max_steps:
+            break
 
-    for step in range(int(max_steps)):
         # Check if we need to start a new epoch
-        epoch_step = step % steps_per_epoch
+        epoch_step = progress.step % steps_per_epoch
         if epoch_step == 0:
-            epoch += 1
-            logger.info(f"Starting epoch {epoch}")
+            progress.epoch += 1
+            logger.info(f"Starting epoch {progress.epoch}")
             # Reshuffle dataset at the beginning of each epoch
-            dataset = dataset.shuffle(seed=(config.seed or 0) + epoch)
+            dataset = dataset.shuffle(seed=(config.seed or 0) + progress.epoch)
 
-        logger.debug(
-            f"Orchestrator step {step} (epoch: {epoch}, epoch_step: {epoch_step}/{steps_per_epoch}, checkpoint step: {ckpt_step})"
-        )
+        logger.debug(f"Orchestrator step {progress.step} ({ckpt_step=})")
         step_start_time = time.time()
 
         # Get the batch
@@ -142,23 +159,23 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
 
         # Optionally, wait for the next checkpoint to be available
         wait_for_weight_ckpt_time, reload_weights_time = 0, 0
-        if step - ckpt_step > config.async_level:
+        if progress.step - ckpt_step > config.async_level:
             logger.debug(
-                f"Hit async barrier because step {step} is {step - ckpt_step} (>{config.async_level}) steps ahead of checkpoint step {ckpt_step}."
+                f"Hit async barrier because step {progress.step} is {progress.step - ckpt_step} (>{config.async_level}) steps ahead of checkpoint step {ckpt_step}."
             )
 
             # Wait for the checkpoint to be available
-            ckpt_step = step - config.async_level
+            ckpt_step = progress.step - config.async_level
             logger.debug(f"Waiting for weight checkpoint for step {ckpt_step}")
             wait_for_weight_ckpt_start_time = time.time()
-            wait_for_weight_checkpoint(config.weights.path, ckpt_step)
+            wait_for_weight_checkpoint(config.weights_path, ckpt_step)
             wait_for_weight_ckpt_time = time.time() - wait_for_weight_ckpt_start_time
             logger.debug(f"Waited {wait_for_weight_ckpt_time:.2f}s for weight checkpoint")
 
             # Reload the weights
             logger.debug(f"Reloading weights for step {ckpt_step}")
             reload_weights_start_time = time.time()
-            await reload_weights(client, config.weights.path, ckpt_step)
+            await reload_weights(client, config.weights_path, ckpt_step)
             reload_weights_time = time.time() - reload_weights_start_time
             logger.debug(f"Reloaded weights in {reload_weights_time:.2f}s")
 
@@ -182,7 +199,7 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
 
         # Get the completions for the batch
         # TODO: Integrate with async (multi-turn) rollout function from verifiers
-        logger.debug(f"Sending {len(batch_messages)} inference requests for step {step}")
+        logger.debug(f"Sending {len(batch_messages)} inference requests for step {progress.step}")
         generate_completions_start_time = time.time()
         # These calls are intentionally non-concurrent because we found that /tokenize is sometimes returning a httpx.ReadError when calling this endpoint concurrently
         input_tokens = [await tokenize(client, config.model, messages) for messages in batch_messages]
@@ -201,7 +218,7 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
 
         # Get the rewards for the completions
         # TODO: Integrate with async scoring function from verifiers
-        logger.debug(f"Computing rewards and advantages for step {step}")
+        logger.debug(f"Computing rewards and advantages for step {progress.step}")
         compute_rewards_start_time = time.time()
         task_types = [problem["task_type"] for problem in problems]
         verification_infos = [json.loads(problem["verification_info"]) for problem in problems]
@@ -216,8 +233,9 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
         num_input_tokens = sum(completion.usage.prompt_tokens for completion in chat_completions)
         num_output_tokens = sum(completion.usage.completion_tokens for completion in chat_completions)
         num_tokens = num_input_tokens + num_output_tokens
-        total_tokens += num_tokens
-        total_samples += config.batch_size
+        progress.total_tokens += num_tokens
+        progress.total_samples += config.batch_size
+        progress.total_problems += config.batch_size // config.sampling.n
         throughput = num_tokens / (generate_completions_time + compute_rewards_time)
         avg_seq_length = num_tokens / config.batch_size
 
@@ -228,7 +246,7 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
                 output_tokens=output_tokens,
                 rewards=rewards,
                 advantages=advantages,
-                step=step,
+                step=progress.step,
             )
 
         # Write serialized batch to disk for trainer workers to consume
@@ -246,27 +264,34 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
             collate_mode=config.collate_mode,
         )
 
-        step_path = Path(config.rollout.path) / f"step_{step}"
+        save_ckpt_time = 0
+        if config.ckpt and config.ckpt.interval and (progress.step + 1) % config.ckpt.interval == 0:
+            logger.debug(f"Saving checkpoint at step {progress.step}")
+            save_ckpt_start_time = time.time()
+            ckpt_manager.save(progress, step=progress.step + 1)
+            save_ckpt_time = time.time() - save_ckpt_start_time
+
+        step_path = Path(config.rollout_path) / f"step_{progress.step}"
         step_path.mkdir(parents=True, exist_ok=True)
         for i, batches in enumerate(all_data_ranks_batches):
             batch_path = step_path / f"rank_{i}.pt"
             tmp_path = batch_path.with_suffix(".tmp")
-            logger.debug(f"Saving rollouts for step {step} for rank {i} to {batch_path}")
+            logger.debug(f"Saving rollouts for step {progress.step} for rank {i} to {batch_path}")
             torch.save(batches, tmp_path)
             tmp_path.rename(batch_path)
 
         # Log step metrics
         step_time = time.time() - step_start_time
-        step_message = f"Orchestrator | step {step} | Time:{step_time:.2f}s | Avg. Reward: {np.mean(rewards):.2f} | Avg. Advantage: {np.mean(advantages):.2f} | Throughput: {throughput:.1f} tokens/s | Avg. Seq. Length: {avg_seq_length:.1f} tokens/sample"
+        step_message = f"Orchestrator | step {progress.step} | Time:{step_time:.2f}s | Avg. Reward: {np.mean(rewards):.2f} | Avg. Advantage: {np.mean(advantages):.2f} | Throughput: {throughput:.1f} tokens/s | Avg. Seq. Length: {avg_seq_length:.1f} tokens/sample"
         logger.info(step_message)
 
         # Log progress metrics to monitor
         progress_metrics = {
-            "progress/infer/total_tokens": total_tokens,
-            "progress/infer/total_samples": total_samples,
-            "progress/train/step": ckpt_step,  # Shared W&B axis
-            "progress/train/epoch": epoch,
-            "step": step,
+            "progress/orchestrator/total_tokens": progress.total_tokens,
+            "progress/orchestrator/total_samples": progress.total_samples,
+            "progress/orchestrator/epoch": progress.epoch,
+            "progress/orchestrator/step": ckpt_step,  # Shared W&B axis
+            "step": progress.step,
         }
         monitor.log(progress_metrics)
 
@@ -274,7 +299,7 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
         perf_metrics = {
             "perf/infer/throughput": throughput,
             "perf/infer/seq_len": avg_seq_length,
-            "step": step,
+            "step": progress.step,
         }
         monitor.log(perf_metrics)
 
@@ -284,21 +309,25 @@ async def orchestrate(config: OrchestratorConfig, setup_queue: Queue | None = No
             "reward/reward_std": np.std(rewards),
             "reward/advantage": np.mean(advantages),
             "reward/advantage_std": np.std(advantages),
-            "step": step,
+            "step": progress.step,
         }
         monitor.log(reward_metrics)
 
         # Log time metrics to monitor
         time_metrics = {
-            "time/infer": step_time,
-            "time/infer/wait_for_weight_ckpt": wait_for_weight_ckpt_time,
-            "time/infer/generate_completions": generate_completions_time,
-            "time/infer/compute_rewards": compute_rewards_time,
-            "time/infer/reload_weights": reload_weights_time,
-            "step": step,
+            "time/orchestrator": step_time,
+            "time/orchestrator/wait_for_weight_ckpt": wait_for_weight_ckpt_time,
+            "time/orchestrator/generate_completions": generate_completions_time,
+            "time/orchestrator/compute_rewards": compute_rewards_time,
+            "time/orchestrator/reload_weights": reload_weights_time,
+            "time/orchestrator/save_ckpt": save_ckpt_time,
+            "step": progress.step,
         }
         monitor.log(time_metrics)
 
+        # Increment progress
+        progress.step += 1
+
     logger.success("Orchestrator finished.")
 
 
diff --git a/src/zeroband/training/orchestrator/utils.py b/src/zeroband/training/orchestrator/utils.py
index 6a988b35..3fe31562 100644
--- a/src/zeroband/training/orchestrator/utils.py
+++ b/src/zeroband/training/orchestrator/utils.py
@@ -48,7 +48,7 @@ def parse_completions(chat_completions: list[ChatCompletion]) -> list[str]:
 
 
 def wait_for_weight_checkpoint(path: Path, step: int, interval: int = 1, log_interval: int = 10) -> None:
-    model_path = Path(path) / f"step_{step}" / "model.pt"
+    model_path = Path(path) / f"step_{step}" / "pytorch_model.bin"
     wait_for_path(model_path, interval, log_interval)
 
 
diff --git a/src/zeroband/training/train.py b/src/zeroband/training/train.py
index 9da0cd19..b42f6e51 100644
--- a/src/zeroband/training/train.py
+++ b/src/zeroband/training/train.py
@@ -5,6 +5,7 @@ import shutil
 import time
 from collections import defaultdict
 from pathlib import Path
+from copy import deepcopy
 
 # Import environment before any other imports
 # ruff: noqa: I001
@@ -15,12 +16,8 @@ import torch
 import torch.distributed as dist
 from torch._guards import log as torch_log
 
-from zeroband.training.ckpt import (
-    TrainingProgress,
-    load_full_checkpoint,
-    save_full_checkpoint,
-    save_weight_checkpoint,
-)
+from zeroband.training.ckpt import CheckpointManager, Progress
+from zeroband.training.weights import save_weight_checkpoint
 from zeroband.training.config import TrainingConfig
 from zeroband.training.data import DataLoader, FakeDataLoader
 from zeroband.training.logger import setup_logger
@@ -82,11 +79,6 @@ def train(config: TrainingConfig):
         else:
             raise RuntimeError(f"Unexpected signal from orchestrator: {signal}")
 
-    # Optionally, clean the checkpoints path
-    if config.ckpt.clean:
-        logger.debug(f"Cleaning checkpoint path {config.ckpt.path}")
-        shutil.rmtree(config.ckpt.path, ignore_errors=True)
-
     # TODO(Mika): Move this to typed env var
     # Allow eager fallback during production so that training runs don't die if compile fails
     if "ZERO_BAND_DEV" not in os.environ:
@@ -110,15 +102,6 @@ def train(config: TrainingConfig):
     model = setup_model(config.model)
     tokenizer = get_tokenizer(config.model)
 
-    # Optionally, initialize a model to compute logprobs
-    if config.recompute_logprobs:
-        logger.info(f"Initializing logprob model ({config.model})")
-        logprob_model = setup_model(config.model)
-
-        # Offload the logprob model to CPU
-        tensor_offloaded_repository: dict[int, OffloadedTensor] = {}
-        tensor_offloaded_repository[0] = offload_model_to_cpu(logprob_model)
-
     # Set up the optimizer
     optimizer = torch.optim.AdamW(
         params=model.parameters(),
@@ -127,11 +110,32 @@ def train(config: TrainingConfig):
         betas=(config.optim.betas1, config.optim.betas2),
     )
 
+    # Get checkpoint manager
+    if config.ckpt:
+        ckpt_manager = CheckpointManager(config.ckpt)
+
     # Optionally, resume training from a checkpoint
-    progress = TrainingProgress()
-    if config.ckpt.resume_path:
-        logger.info(f"Resuming training from checkpoint {config.ckpt.resume_path}")
-        load_full_checkpoint(model, [optimizer], progress, config.ckpt.resume_path)
+    progress = Progress()
+    if config.ckpt and config.ckpt.resume_step:
+        logger.info(f"Resuming training from checkpoint step `{config.ckpt.resume_step}`")
+        ckpt_manager.load(model, [optimizer], progress, step=config.ckpt.resume_step)
+
+    # Optionally, initialize a model to compute logprobs
+    if config.recompute_logprobs:
+        # Offload the logprob model to CPU
+        tensor_offloaded_repository: dict[int, OffloadedTensor] = {}
+        infer_step = max(progress.step - config.async_level, 0)
+        for async_step in range(infer_step, progress.step + 1):
+            logger.info(f"Initializing logprob model ({config.model}) for step {async_step}")
+            model_name_or_path = (
+                config.model.name
+                if not (config.ckpt and config.ckpt.resume_step)
+                else config.weights.path / f"step_{async_step}"
+            )
+            model_config = deepcopy(config.model)
+            model_config.name = model_name_or_path
+            logprob_model = setup_model(model_config)
+            tensor_offloaded_repository[async_step] = offload_model_to_cpu(logprob_model)
 
     # Set up the data loader (Optionally, use a fake data loader for debugging)
     logger.info(f"Initializing data loader ({config.data})")
@@ -248,7 +252,6 @@ def train(config: TrainingConfig):
             importance_ratio = importance_ratio / num_micro_batches
 
             # Backward pass (ensures loss reduction across FSDP ranks)
-            logger.debug(f"Backward pass on micro batch {micro_step} / {num_micro_batches}")
             loss.backward()
 
             loss_metrics["loss/loss"] += loss.detach().clone()
@@ -282,15 +285,17 @@ def train(config: TrainingConfig):
             logger.info(f"Broadcasting weights from {model_path} via shardcast")
             shardcast.broadcast(model_path.as_posix())  # TODO: Is this blocking?
 
-        # Optionally, remove old weight checkpoints to save space
-        # +1 to ensure to not delete current checkpoint when async_level=0
-        if len(active_weight_checkpoint_paths) > config.async_level + 1:
-            path_to_delete = active_weight_checkpoint_paths.pop(0)
-            ckpt_step = int(path_to_delete.name.split("_")[-1])
-            should_keep = config.weights.interval and ckpt_step % config.weights.interval == 0
-            if not should_keep:
-                logger.debug(f"Removing past weight checkpoint ({path_to_delete})")
-                shutil.rmtree(path_to_delete, ignore_errors=True)
+        # Consider cleaning up weight ckpt from async_level+1 steps ago
+        candidate_path_to_delete = Path(config.weights.path) / f"step_{max(progress.step - (config.async_level + 1), 0)}"
+        logger.debug(f"Considering to delete weight checkpoint {candidate_path_to_delete}")
+        candidate_weight_step_to_delete = int(candidate_path_to_delete.name.split("_")[-1])
+        keep_for_eval = config.weights.interval and candidate_weight_step_to_delete % config.weights.interval == 0
+        # For checkpointing step x, we need all weight checkpoints in [x-async_level, x] (for logprob model)
+        # To get [n-k, n] with interval n and buffer k over all natural numbers x, we use the condition (n - (x % n)) % n <= k
+        keep_for_ckpt = config.ckpt and (config.ckpt.interval - (candidate_weight_step_to_delete % config.ckpt.interval)) % config.ckpt.interval <= config.async_level 
+        if not (keep_for_eval or keep_for_ckpt):
+            logger.debug(f"Removing past weight checkpoint {candidate_path_to_delete} ({keep_for_eval=}, {keep_for_ckpt=})")
+            shutil.rmtree(candidate_path_to_delete, ignore_errors=True)
 
         # Optionally, dump memory snapshot
         if config.profile_path and progress.step == 2 and world.rank == 0:
@@ -301,14 +306,6 @@ def train(config: TrainingConfig):
             torch.cuda.memory._dump_snapshot(profile_path.as_posix())
             torch.cuda.memory._record_memory_history(enabled=False)
 
-        # Optionally, save the full checkpoint
-        save_ckpt_time = 0
-        if config.ckpt and config.ckpt.interval and progress.step % config.ckpt.interval == 0:
-            logger.debug(f"Saving checkpoint at step {progress.step}")
-            save_ckpt_start_time = time.time()
-            save_full_checkpoint(model, [optimizer], progress, config.ckpt.path)
-            save_ckpt_time = time.time() - save_ckpt_start_time
-
         # Update the CPU logprob model to updated model
         if config.recompute_logprobs:
             logger.debug("Offloading updated model to CPU")
@@ -327,6 +324,16 @@ def train(config: TrainingConfig):
         mfu = perf_counter.get_mfu() or 0
         loss_metrics = {key: value.item() for key, value in loss_metrics.items()}
 
+        # Optionally, save the full checkpoint
+        # We save the checkpoint here, because the total_tokens and total_samples are updated yet
+        # But because the step is only incremented at the end of the step, we need to increment it manually for the checkpoint
+        save_ckpt_time = 0
+        if config.ckpt and config.ckpt.interval and (progress.step + 1) % config.ckpt.interval == 0:
+            logger.debug(f"Saving checkpoint at step {progress.step}")
+            save_ckpt_start_time = time.time()
+            ckpt_manager.save(model, [optimizer], progress, step=progress.step + 1)
+            save_ckpt_time = time.time() - save_ckpt_start_time
+
         # Log step metrics
         step_time = time.time() - step_start_time
         step_message = f"Training     | step {progress.step} | Time:{step_time:.2f}s | Loss: {loss_metrics['loss/loss']:.2f} | Entropy: {loss_metrics['loss/entropy']:.2f} | Mean Ratio: {loss_metrics['loss/importance_ratio']:.2f} | {throughput:.0f} tokens/s | MFU: {mfu:.1f}%"
diff --git a/src/zeroband/training/weights.py b/src/zeroband/training/weights.py
new file mode 100644
index 00000000..664e1274
--- /dev/null
+++ b/src/zeroband/training/weights.py
@@ -0,0 +1,82 @@
+import threading
+import time
+from pathlib import Path
+
+import torch
+from torch.distributed.checkpoint.state_dict import _get_fqns as get_fqns
+from torch.distributed.tensor import DTensor
+from transformers import AutoTokenizer
+
+from zeroband.training.model import Model
+from zeroband.training.world import get_world
+from zeroband.utils.logger import get_logger
+
+
+def save_weight_checkpoint(
+    model: Model,
+    tokenizer: AutoTokenizer,
+    path: Path,
+    dtype: torch.dtype = torch.bfloat16,
+    async_save: bool = False,
+) -> Path:
+    """
+    Save a HF-compatible weight-only checkpoint to the specified path. Saves the
+    model weights as `pytorch_model.bin`, the model config, generation arguments and tokenizer
+    using HF's `save_pretrained` method.
+    """
+    # Get logger and world info
+    logger = get_logger()
+    is_master = get_world().rank == 0
+
+    # Create checkpoint directory if it doesn't exist
+    if not path.exists():
+        path.mkdir(parents=True, exist_ok=True)
+
+    # Gather distributed weights for weight checkpoint
+    start_time = time.time()
+    logger.debug("Gathering sharded weights")
+    cpu_state = {}
+    for key, value in model.state_dict().items():
+        if isinstance(value, DTensor):
+            value = value.to(dtype)
+            # only gather after the downcast to dtype as it will be faster
+            value = value.full_tensor()
+
+        if is_master:
+            key = get_fqns(model, key)
+            assert len(key) == 1
+            key = next(iter(key))
+            # TODO(Sami) Blocking to avoid race condition, should make non-blocking long-term tho
+            cpu_state[key] = value.to("cpu", non_blocking=False)
+
+    torch.distributed.barrier()
+    logger.debug(f"Gathered sharded weights in {time.time() - start_time:.2f} seconds")
+
+    model_path = path / "pytorch_model.bin"
+
+    def _save_weight_checkpoint():
+        if is_master:
+            start_time = time.time()
+            logger.debug(f"Saving weights to {path}")
+
+            # Save model weights to temporary file to avoid race condition
+            tmp_model_path = model_path.with_suffix(".tmp")
+            torch.save(cpu_state, tmp_model_path)
+
+            # Rename temporary file to indicate checkpoint is complete
+            tmp_model_path.rename(model_path)
+
+            # Save model config, generation arguments and tokenizer
+            model.config.save_pretrained(path)
+            model.generation_config.save_pretrained(path)
+            tokenizer.save_pretrained(path)
+
+            logger.debug(f"Saved weights to {path} in {time.time() - start_time:.2f} seconds")
+
+    if async_save:
+        thread = threading.Thread(target=_save_weight_checkpoint)
+        thread.start()
+    else:
+        _save_weight_checkpoint()
+
+    return model_path
diff --git a/src/zeroband/utils/config.py b/src/zeroband/utils/config.py
index e1d2d4f1..a25c419f 100644
--- a/src/zeroband/utils/config.py
+++ b/src/zeroband/utils/config.py
@@ -94,6 +94,13 @@ class WandbMonitorConfig(BaseConfig):
         ),
     ] = None
 
+    id: Annotated[
+        str | None,
+        Field(
+            description="The W&B run ID to log to. If None, a random ID will be generated. If you want to resume a run, you can set the ID to the run ID you want to resume.",
+        ),
+    ] = None
+
     dir: Annotated[
         Path | None,
         Field(
diff --git a/src/zeroband/utils/monitor.py b/src/zeroband/utils/monitor.py
index 1b748b0c..0e6786c9 100644
--- a/src/zeroband/utils/monitor.py
+++ b/src/zeroband/utils/monitor.py
@@ -137,7 +137,9 @@ class WandbMonitor(Monitor):
             project=config.project,
             group=config.group,
             name=config.name,
+            id=config.id,
             dir=config.dir,
+            resume="allow",
             config=run_config.model_dump() if run_config else None,
             mode="offline" if config.offline else None,
         )
